### PR TITLE
List and delete policy renewals with gql or when renewing policy

### PR DESCRIPTION
**Pull request by toch**
List and delete policy renewals with gql or when renewing policy
    
Policy renewals were directly exposed by the C# REST API. The policy mobile app
needs to list, delete, and add policy renewals. The GraphQL API didn't allow
to do so.
    
With the present change, we can now list and delete policy renewals with a
GrapQL query and mutation (resp.). As there is already a mutation to renew
a given policy, we haven't added a dedicated GraphQL mutation to add
a policy renewal independently. As there is a background task scheduling policy
renewals, we are now soft deleting the renewals of a given policy when we                                                                     
manually renew it.

Open Questions:

* how would you test that? I don't see existing suite testing the gql queries and mutations
* when renewing a policy, it is checked if it's idle or not. Once renewed, its status stays the same. Is it normal? shouldn't it change?

### CODE REVIEW

File: policy/schema.py [hunk 1]
```diff
@@ -338,6 +345,20 @@ def on_policy_mutation(sender, **kwargs):
...
+def on_policy_renewal_mutation(sender, **kwargs):
+    uuids = kwargs['data'].get('uuids', [])
+    if not uuids:
+        uuid = kwargs['data'].get('policy_renewal_uuid', None)
+        uuids = [uuid] if uuid else []
+    if not uuids:
+        return []
+    impacted_policy_renewals = PolicyRenewal.objects.filter(uuid__in=uuids).all()
+    for policy_renewal in impacted_policy_renewals:
+        PolicyRenewalMutation.objects.create(
```
Comment (edarchis) [inline policy/schema.py:357]: There is an object_mutated helper for this but this works too.

File: policy/services.py [hunk 2]
```diff
@@ -37,6 +37,27 @@ def reset_policy_before_update(policy):
...
 
 
+class PolicyRenewalService:
+    def __init__(self, user):
+        self.user = user
+
+    def delete(self, policy_renewal):
+        try:
+            policy_renewal = PolicyRenewal.objects.get(uuid=policy_renewal.uuid)
+            logger.debug(f'YEAHHHH {policy_renewal.uuid}')
```
Comment (edarchis) [inline policy/services.py:47]: Maybe temper your enthusiasm a bit :)

File: policy/gql_mutations.py [hunk 3]
```diff
@@ -151,6 +153,39 @@ def async_mutate(cls, user, **data):
...
+class DeletePolicyRenewalsMutation(OpenIMISMutation):
+    _mutation_module = "policy"
+    _mutation_class = "DeletePolicyRenewalsMutation"
+
+    class Input(OpenIMISMutation.Input):
+        uuids = graphene.List(graphene.String)
+
+    @classmethod
+    def async_mutate(cls, user, **data):
+        try:
```
Comment (edarchis) [inline policy/gql_mutations.py:165]: Shouldn't we validate the logged user/officer ?

File: policy/services.py [hunk 4]
```diff
@@ -37,6 +37,27 @@ def reset_policy_before_update(policy):
     policy.officer_id = None
 
 
+class PolicyRenewalService:
+    def __init__(self, user):
+        self.user = user
+
+    def delete(self, policy_renewal):
+        try:
+            policy_renewal = PolicyRenewal.objects.get(uuid=policy_renewal.uuid)
```
Comment (tdethier) [inline policy/services.py:46]: Maybe it's a dumb question, but why do you search for the `policy_renewal` instance if you already receive it as a parameter? Or maybe these are 2 different things with the same name?

File: policy/schema.py (hunk 1 above)
Comment (toch) [reply policy/schema.py:357]: should I use it as it is at `gql_mutations.py#CreateRenewOrUpdatePolicyMutation.do_mutate#L51`?

```python
PolicyMutation.object_mutated(user, client_mutation_id=client_mutation_id, policy=policy)
```

File: policy/services.py (hunk 2 above)
Comment (toch) [reply policy/services.py:47]: thanks!

File: policy/services.py (hunk 4 above)
Comment (tdethier) [reply policy/services.py:46]: Yes!

**Comment by tdethier** [comment]
This is no longer needed, we'll prepare a new PR instead

**Pull request closed without merging**